### PR TITLE
Improve performances of batch pathfinding and simulation :rocket: 

### LINKED ISSUE #7913: gateway: timeout on a timetable with 1410 trains
### What happened?

Go on this [timetable](https://rec-osrd.reseau.sncf.fr/operational-studies/projects/4/studies/52/scenarios/133). The gateway timeout for path projection.

image
image

### What did you expect to happen?

The gateway to wait for an answer or editoast to be ready to answer faster.

### How can we reproduce it (as minimally and precisely as possible)?

1. open this [scenario](https://rec-osrd.reseau.sncf.fr/operational-studies/projects/4/studies/52/scenarios/133)

### What operating system, browser and environment are you using?

- **Browser**: Firefox 

### OSRD version (top right corner `Account` button > `Informations`)

60ae156

### PR DESCRIPTION
**Pull request by flomonster**
## What is done

- Massive reduction of Redis call
- Stop handling the expiration time of Redis entries

close #7913

## Benchmark

We are testing two endpoints using batch simulation and pathfinding:
- Simulation summary
- Project path

Tests are performed with everything cached **locally**. I'm guessing that the gain will be much greater in prod with network overhead.

### Before

- Simulation summary: 1120ms
- Project path: 1400ms

image

### After

- Simulation summary: 800ms
- Project path: 950ms

image

### CODE REVIEW

File: editoast/src/redis_utils.rs [hunk 1]
```diff
@@ -209,41 +188,3 @@ impl RedisClient {
...
-#[cfg(test)]
-mod tests {
-    use super::RedisClient;
-    use rstest::rstest;
-    use serde::{Deserialize, Serialize};
-    use tokio::time::{sleep, Duration};
-
-    /// Test get and set json values to redis
-    #[rstest]
-    async fn json_get_set() {
```
Comment (woshilapin) [inline editoast/src/redis_utils.rs:222]: Any reason to remove this test? I see that we might need to modify it because of the expiration. We might even want to expand that test suite, since we don't test most of the edge cases?

File: editoast/src/views/v2/path/pathfinding.rs [hunk 2]
```diff
@@ -95,117 +94,189 @@ pub async fn post(
...
-    if path_items.len() <= 1 {
-        return Ok(PathfindingResult::NotEnoughPathItems);
+    let mut to_cache = vec![];
+    let mut pathfinding_requests = vec![];
+    let mut pathfinding_requests_index = vec![];
+    for (index, (result, path_input)) in results.iter_mut().zip(path_inputs).enumerate() {
+        if result.is_some() {
+            continue;
+        }
+        let path_items = path_input.clone().path_items;
```
Comment (woshilapin) [inline editoast/src/views/v2/path/pathfinding.rs]: I don't believe you have to clone here.

```suggestion
        let path_items = &path_input.path_items;
```

**Review by Wadjetz — APPROVED** [review]

File: editoast/src/views/v2/path/pathfinding.rs [hunk 3]
```diff
@@ -95,117 +94,201 @@ pub async fn post(
...
+    let mut pathfinding_requests_index = vec![];
+    for (index, (pathfinding_result, pathfinding_input)) in pathfinding_results
+        .iter_mut()
+        .zip(pathfinding_inputs)
+        .enumerate()
     {
-        return Ok(pathfinding_result);
+        if pathfinding_result.is_some() {
+            continue;
+        }
```
Comment (Castavo) [inline editoast/src/views/v2/path/pathfinding.rs:153]: Could this be moved in a build_pathfinding_request function, return a Result<PathfindingRequest, PathfindingResult> ?
The boiler plate of adding stuff to cache and updating patfinding result would be factorized

File: editoast/src/views/v2/path/pathfinding.rs [hunk 4]
```diff
@@ -95,117 +94,201 @@ pub async fn post(
...
+}
+
+/// Compute a path given a batch of trainschedule and an infrastructure.
+pub async fn pathfinding_from_train_batch(
+    conn: &mut DbConnection,
+    redis: &mut RedisConnection,
+    core: Arc<CoreClient>,
+    infra: &Infra,
+    train_schedules: &[TrainSchedule],
+    rolling_stocks: &HashMap<String, RollingStockModel>,
```
Comment (Castavo) [inline editoast/src/views/v2/path/pathfinding.rs:254]: nit: it looks cumbersome to handle a separate mapping of rolling stocks, maybe we could introduce some sort of "FullTrainSchedule" struct, which would hold its rolling stock ?

Comment (flomonster) [reply editoast/src/views/v2/path/pathfinding.rs:254]: Good idea but I think I'll do it later.

> [!NOTE]
> We might want to store a reference and not the whole rolling stock to avoid duplication.

File: editoast/src/views/v2/path/pathfinding.rs (hunk 3 above)
Comment (Castavo) [reply editoast/src/views/v2/path/pathfinding.rs:153]: I messed up my diff selection for this comment, it relates to the body of the main for loop of `pathfinding_blocks_batch` (the one surrounding this comment)

Comment (flomonster) [reply editoast/src/views/v2/path/pathfinding.rs:153]: Good catch.

**Review by woshilapin — APPROVED** [review]
Nice improvement
